policy_agent/policy_agent.py

```
@@ -0,0 +1,643 @@
+                self.endpoints[ep_key] = ep
+                del self.changed_endpoints[ep_key]
+
+            # else define service policy
+            else:
+                # find pods and append profiles
+                associations = ep.generate_svc_profiles_pods()
+                for profile, pods in associations.items():
+                    for pod in pods:
+                        existing_pod = self.match_pod(
+                            namespace=ep.namespace, name=pod)
+                        #_log.info("Adding profile %s to pod %s" % (profile, existing_pod))
+                        if existing_pod:
+                            try:
+                                _datastore_client.append_profiles_to_endpoint(profile_names=[profile],
+                                                                              endpoint_id=existing_pod.ep_id)
+                            except ProfileAlreadyInEndpoint:
+                                _log.warning(
+                                    "Applying %s to Pod %s : Profile Already exists" % (profile, existing_pod.key))
+                        else:
+                            _log.warning("Pod %s is not yet processed" % (pod))
+
+                # declare Endpoints obj processed
+                self.epRV = ep.RV
+                self.endpoints[ep_key] = ep
+                del self.changed_endpoints[ep_key]
+
+
+class Resource():
+
+    """
+    Resource objects pull pertinent info from json blobs and maintain universal functions
+    """
+
+    def __init__(self, json):
+        """
+        On init, each Resource saves the raw json, pulls necessary info (unique),
+        and defines a unique key identifier
+        """
+        self.json = json
+        self.RV = json["metadata"]["resourceVersion"]
+        self.from_json(json)
+        self.key = self.get_key()
+
+    def from_json(self, json):
+        self.name = "noop"
+        return
+
+    def get_key(self):
```

This is meant to be overridden - any subclass of "Resource" must override this method.

Instead of providing a default, this should raise `NotImplementedError` to indicate that the subclass hasn't overridden this method.